### PR TITLE
Scala: Ensure implicit conversion keeps type fidelity

### PR DESCRIPTION
**Pull request by rozza**
When converting from a BsonDocument to org.bson.Document
the scala driver previously used json. The implementation
should not rely on the json output format, but rather use
the Codecs to do the conversion.

JAVA-3916

### CODE REVIEW

File: driver-scala/src/test/scala/org/mongodb/scala/ScalaPackageSpec.scala [hunk 1]
```diff
@@ -141,4 +140,33 @@ class ScalaPackageSpec extends BaseSpec {
     val javaCredential5 = JMongoCredential.createGSSAPICredential("userName")
     scalaCredential5 should equal(javaCredential5)
   }
+
+  it should "implicitly convert to org.Bson.document with type fidelity" in {
```
Comment (jyemin) [inline driver-scala/src/test/scala/org/mongodb/scala/ScalaPackageSpec.scala:144]: Lowercase the `B`

File: driver-scala/src/main/scala/org/mongodb/scala/package.scala [hunk 2]
```diff
@@ -392,9 +399,18 @@ package object scala extends ClientSessionImplicits with ObservableImplicits wit
 
   implicit def bsonDocumentToDocument(doc: BsonDocument): Document = new Document(doc)
 
-  implicit def bsonDocumentToUntypedDocument(doc: BsonDocument): org.bson.Document =
-    org.bson.Document.parse(doc.toJson())
+  implicit def documentToUntypedDocument(doc: Document): org.bson.Document =
+    bsonDocumentToUntypedDocument(doc.underlying)
+
+  implicit def bsonDocumentToUntypedDocument(doc: BsonDocument): org.bson.Document = {
+    val writer: BsonBinaryWriter = new BsonBinaryWriter(new BasicOutputBuffer())
```
Comment (jyemin) [inline driver-scala/src/main/scala/org/mongodb/scala/package.scala:406]: You can use BsonDocumentReader and simplify to

    new DocumentCodec().decode(new BsonDocumentReader(doc), DecoderContext.builder().build())

**Review by jyemin — CHANGES_REQUESTED** [review]

File: driver-scala/src/main/scala/org/mongodb/scala/package.scala (hunk 2 above)
Comment (rozza) [reply driver-scala/src/main/scala/org/mongodb/scala/package.scala:406]: Much nicer thanks

File: driver-scala/src/main/scala/org/mongodb/scala/package.scala [hunk 3]
```diff
@@ -392,9 +399,18 @@ package object scala extends ClientSessionImplicits with ObservableImplicits wit
...
+    new BsonDocumentCodec().encode(writer, doc, EncoderContext.builder().build())
```
Comment (stIncMale) [inline driver-scala/src/main/scala/org/mongodb/scala/package.scala:407]: I am not sure, but at first glance`BsonDocumentCodec` looks like an immutable class. If that is the case, do we actually want to create a new object per the `bsonDocumentToUntypedDocument` method invocation? Unless we are fortunate and [scalar replacement](https://shipilev.net/jvm/anatomy-quarks/18-scalar-replacement/) kicks in, this creates unnecessary garbage.

Comment (rozza) [reply driver-scala/src/main/scala/org/mongodb/scala/package.scala:407]: Good call - done.

**Review by jyemin — APPROVED** [review]

File: driver-scala/src/test/scala/org/mongodb/scala/ScalaPackageSpec.scala [hunk 4]
```diff
@@ -141,4 +140,33 @@ class ScalaPackageSpec extends BaseSpec {
     val javaCredential5 = JMongoCredential.createGSSAPICredential("userName")
     scalaCredential5 should equal(javaCredential5)
   }
+
+  it should "implicitly convert to org.bson.document with type fidelity" in {
+
+    val bsonDocument = Document(
+      "null" -> BsonNull(),
+      "int32" -> BsonInt32(32),
+      "int64" -> BsonInt64(64L),
```
Comment (aaabramov) [inline driver-scala/src/test/scala/org/mongodb/scala/ScalaPackageSpec.scala:149]: I'd change `64L` -> `Long.MaxValue` for reproducing issue in JIRA ticket.

FYI, I am OP of the bug

Comment (rozza) [reply driver-scala/src/test/scala/org/mongodb/scala/ScalaPackageSpec.scala:149]: Sounds good

File: driver-scala/src/test/scala/org/mongodb/scala/ScalaPackageSpec.scala [hunk 5]
```diff
@@ -141,4 +140,33 @@ class ScalaPackageSpec extends BaseSpec {
...
     scalaCredential5 should equal(javaCredential5)
   }
+
+  it should "implicitly convert to org.bson.document with type fidelity" in {
+
+    val bsonDocument = Document(
+      "null" -> BsonNull(),
+      "int32" -> BsonInt32(32),
+      "int64" -> BsonInt64(Long.MaxValue),
+      "decimal128" -> BsonDecimal128(128),
```
Comment (aaabramov) [inline driver-scala/src/test/scala/org/mongodb/scala/ScalaPackageSpec.scala:150]: Also, I would explicitly pass 128.1 as with BsonDouble.

**Pull request merged**